classes/choice_importer.php

```diff
@@ -0,0 +1,305 @@
+     */
+    public function import($content, $live=true) {
+        global $DB;
+
+        $reader = $this->get_reader();
+
+        $importstatus = new \stdClass;
+        $importstatus->status = self::IMPORT_STATUS_OK;  // Unless we hear otherwise.
+        $importstatus->live = $live;  // Only commit live transactions.
+        $importstatus->errors = array();
+        $importstatus->importcount = 0;
+        $importstatus->rowcount = 1;  // Start at 1 for header.
+
+        $importstatus->readcount = $reader->load_csv_content($content, 'UTF-8', 'comma');
+        $importerror = $reader->get_error();
+
+        if ($importerror) {
+            $importstatus->status = self::IMPORT_STATUS_SETUP_ERROR;
+            $importstatus->errors[] = $importerror;
+        } else {
+            if (empty($importstatus->readcount)) {
+                $importstatus->status = self::IMPORT_STATUS_DATA_ERROR;
+                $importstatus->errors[] = get_string('csvempty', 'ratingallocate');
+            } else {
+                // Trim whitespace on headers, just in case.
+                if (!$fieldnames = array_map('trim', $reader->get_columns())) {
+                    throw new moodle_exception('cannotreadtmpfile', 'error');
+                }
+
+                // Check for any missing required fields.
+                $missingfields = array_diff(self::REQUIRED_FIELDS, $fieldnames);
+                if ($missingfields) {
+                    $importstatus->status = self::IMPORT_STATUS_SETUP_ERROR;
+                    $importstatus->errors[] = get_string('csvupload_missing_fields', 'ratingallocate',
+                        self::print_fields($missingfields));
+                    return $importstatus;
+                }
+
+                // Map group names to group IDs.
+                $allgroups = $this->ratingallocate->get_group_selections();
+                $groupidmap = array_flip($allgroups);
+
+                // Start DB transaction.
+                $transaction = $this->ratingallocate->db->start_delegated_transaction();
+                try {
+
+                    $reader->init();
+                    while ($record = $reader->next()) {
+                        $importstatus->rowcount++;
+                        $recordmap = new \stdClass();
+
+                        // Map cell contents to field names.
+                        foreach ($record as $col => $cell) {
+                            $fieldname = $fieldnames[$col];
+
+                            // Skip non-required field columns.
+                            if (!in_array($fieldname, self::REQUIRED_FIELDS)) {
+                                continue;
+                            }
+
+                            if ($fieldname == 'groups') {
+                                $groups = array();
+
+                                // Turn off 'usegroups' if no groups specified.
+                                if (empty($cell)) {
```

Comment:
This should be trimmed for whitespaces first, otherwise a trailing whitespace will lead to a choice with usegroups enabled without an assigned group.